### PR TITLE
Ensure main query loop is entered for singular content in block themes

### PR DESCRIPTION
**Pull request by felixarntz**
While initially it seemed that this had some overlap with https://github.com/WordPress/gutenberg/pull/49904, a closer review of the underlying issues determined this problem being separate: The other PR addresses problems with the main query loop _in general_, while this PR and the below ticket only relate to singular templates in block themes and that they commonly lack the necessary blocks `core/query` and `core/post-template` which are used to start the main query loop.

This PR ensures that the main query loop is still started for block theme templates that lack those blocks. While alternatively we could consider introducing a new "invisible" wrapper block for this purpose, that would then require pretty much all block themes out there to adjust their templates, needless to say that any templates already modified in the database would not receive those updates. So while the fix in this PR is a workaround, it is reliable as there is by definition only a single post in the main query for singular content and thus we can wrap the entire template into the main query loop.

Trac ticket: https://core.trac.wordpress.org/ticket/58154

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

### CODE REVIEW

File: src/wp-includes/block-template.php [hunk 1]
```diff
@@ -228,7 +228,25 @@ function get_the_block_template_html() {
...
+	/*
+	 * Most block themes omit the `core/query` and `core/post-template` blocks in their singular content templates.
+	 * While this technically still works since singular content templates are always for only one post, it results in
+	 * the main query loop never being entered which causes bugs in core and the plugin ecosystem.
+	 *
+	 * The workaround below ensures that the loop is started even for those singular templates without the relevant
+	 * blocks. The while loop will by definition only go through a single iteration, i.e. `do_blocks()` is only called
+	 * once.
+	 */
+	if ( is_singular() && ! has_block( 'core/post-template', $content ) ) {
```
Comment (youknowriad) [inline src/wp-includes/block-template.php]: I don't understand why we're checking the 'core/post-template' block's absence here? Can you clarify?

Comment (felixarntz) [reply src/wp-includes/block-template.php]: My rationale for adding it was that a block theme author may still add the block to their singular template within a `core/query` block with `inherit: true`, which would trigger the main query loop anyway - which would still be slightly cleaner so I was thinking in that case we should respect that and use that approach to trigger the loop rather than globally doing it here. LMKWYT.

Comment (youknowriad) [reply src/wp-includes/block-template.php]: I think the main use-case for query (and post template) within singular templates is without "inherit: true" for things like "related posts with similar categories" or "recent posts" or things like that. So I think this check is probably going to reintroduce the "error" for most of these themes.

My question now becomes: what happens if we drop the check and we end up running the query twice: one here and one with the `core/post-template` block (with inherit true)?

Comment (felixarntz) [reply src/wp-includes/block-template.php]: > I think the main use-case for query (and post template) within singular templates is without "inherit: true" for things like "related posts with similar categories" or "recent posts" or things like that. So I think this check is probably going to reintroduce the "error" for most of these themes.

Makes sense. We should either remove this, or we would need to implement a more precise check that only triggers the loop if no `core/post-template` block wrapped by a `core/query` block with `inherit: true` is present - but that may be quite complicated to implement reasonably within this logic.

> My question now becomes: what happens if we drop the check and we end up running the query twice: one here and one with the `core/post-template` block (with inherit true)?

In that case, because of the fix in https://github.com/WordPress/gutenberg/pull/49904, it would just run through a "duplicated" loop within the main query loop, which would be fine though since it would be just for the same single post. Basically it would be something like this:
```php
// Main query loop is only run once for the single post.
while ( have_posts() ) {
	the_post();

	// Copy of main query loop from `core/post-template` is only run once for the same post.
	while ( $duplicated_query->have_posts() ) {
		$duplicated_query->the_post();

		// Display the post.
	}
}
```

So while that feels a bit odd, there wouldn't be an actual problem, since the duplicated loop would still be within the main query loop, so things would work as expected.

Based on that, I should probably remove the check for the `core/post-template` block.

Comment (felixarntz) [reply src/wp-includes/block-template.php]: Updated in https://github.com/WordPress/wordpress-develop/pull/5104/commits/0726f9b85cd4f6b6ae616bb93994d48f5b4980ff.

**Comment by youknowriad** [comment]
This looks good to me. Is this dependent on the Gutenberg fix? Should we be waiting for the package update first before committing?

**Comment by felixarntz** [comment]
Thanks @youknowriad!
> This looks good to me. Is this dependent on the Gutenberg fix? Should we be waiting for the package update first before committing?

No, in fact it would be better to first commit this one before backporting the other Gutenberg fix, see https://core.trac.wordpress.org/ticket/58154#comment:15.

File: src/wp-includes/block-template.php [hunk 2]
```diff
@@ -228,7 +228,28 @@ function get_the_block_template_html() {
...
+	 *
+	 * The workaround below ensures that the loop is started even for those singular templates. The while loop will by
+	 * definition only go through a single iteration, i.e. `do_blocks()` is only called once.
+	 *
+	 * Even if the block template contained a `core/query` and `core/post-template` block referencing the main query
+	 * loop, it would not cause errors since it would use a cloned instance and go through the same loop of a single
+	 * post, within the actual main query loop.
+	 */
+	if ( is_singular() ) {
+		while ( have_posts() ) {
```
Comment (joemcgill) [inline src/wp-includes/block-template.php:247]: Since we're expecting this to only run once anyway, would it be safer for us to use an `if` here? Also, is there a possibility that a singular template could load in a bizarre scenario where `have_posts()` would return `false` and then no content is returned? I can't think of an edge case where that would occur, but seems technically possible if someone has modified the main query somehow.

**Review by joemcgill — APPROVED** [review]
This change makes sense to me. Left some questions inline as more of a thinking exercise than something that I think necessarily needs to be addressed.

File: src/wp-includes/block-template.php (hunk 2 above)
Comment (felixarntz) [reply src/wp-includes/block-template.php:247]: > Since we're expecting this to only run once anyway, would it be safer for us to use an `if` here?

We could do that, however then we would still also need to make the 2nd `have_posts()` call which would by definition return `false` but also run the logic to close the loop. Basically this:
```php
if ( is_singular() && have_posts() ) {
	the_post();
	$content = do_blocks( $content );
	have_posts(); // Closes main query loop.
} else {
	$content = do_blocks( $content );
}
```

> Also, is there a possibility that a singular template could load in a bizarre scenario where `have_posts()` would return `false` and then no content is returned? I can't think of an edge case where that would occur, but seems technically possible if someone has modified the main query somehow.

Not unless someone intentionally wanted to break WordPress. I think by not using the `while()`, we should be safe here, since even if someone tampered with the main query in (clearly wrong!) ways, it would now only parse the first post. If we wanted to add yet another check to be double-sure, maybe we could change the first if clause from my above code to `is_singular() && 1 === $wp_query->post_count && have_posts()`.

WDYT?

Comment (joemcgill) [reply src/wp-includes/block-template.php:247]: Ah, I didn't remember that the final `have_posts()` call is what also sets `in_the_loop` to `false` ([among other things](https://github.com/WordPress/wordpress-develop/blob/6.3/src/wp-includes/class-wp-query.php#L3675-L3704)). In that case, leaving it as you currently has it makes more sense. 👍🏻

File: src/wp-includes/block-template.php [hunk 3]
```diff
@@ -228,7 +228,28 @@ function get_the_block_template_html() {
...
+	 * the main query loop never being entered which causes bugs in core and the plugin ecosystem.
+	 *
+	 * The workaround below ensures that the loop is started even for those singular templates. The while loop will by
+	 * definition only go through a single iteration, i.e. `do_blocks()` is only called once.
+	 *
+	 * Even if the block template contained a `core/query` and `core/post-template` block referencing the main query
+	 * loop, it would not cause errors since it would use a cloned instance and go through the same loop of a single
+	 * post, within the actual main query loop.
+	 */
+	if ( is_singular() ) {
```
Comment (felixarntz) [inline src/wp-includes/block-template.php]: @joemcgill Based on your other comment, what do you think about adding this? IMO it would make this code a bit safer, as it would ensure that, _if_ something weird is going on and the main query was tampered with in a way that would change the number of posts in the result, we would fall back to the old code (`else` clause).
```suggestion
	global $wp_query;
	if ( is_singular() && 1 === $wp_query->post_count && have_posts() ) {
```

File: src/wp-includes/block-template.php (hunk 2 above)
Comment (felixarntz) [reply src/wp-includes/block-template.php:247]: SG regarding the `while`, however I left another comment in https://github.com/WordPress/wordpress-develop/pull/5104#discussion_r1309421673 regarding the other thing mentioned above.

**Comment by gziolo** [comment]
I did quite extensive testing using the Gutenberg plugin version downloaded from the linked PR:

https://github.com/WordPress/gutenberg/actions/runs/6003053680

https://github.com/WordPress/gutenberg/suites/15571281198/artifacts/888604674

I identified some edge cases in the block-based theme that I believe we saw together with @felixarntz last week during WC US:

**Homepage rendered on the front**

Screenshot 2023-08-30 at 15 19 16

Everything works here as expected – "Third post with loop" renders the same loop as the one on the homepage and it correctly skips rendering the post content, which would lead to infinite loop.

**Blog Home template in the editor**

Screenshot 2023-08-30 at 15 23 18

For some reasons, the nested query loop doesn't get displayed, but that might be good in this context.

**Third post with loop on the frontend**

Screenshot 2023-08-30 at 15 26 22

The debugging message is only visible when WP is in the debug mode. More importantly, the query loop isn't correctly rendered.

 **Third post with loop in the editor**

Screenshot 2023-08-30 at 15 25 46

Everything works as expected in the editor though.

---

One more thing that I found surprising is that the featured image from the first post gets rendered on top of the homepage:

Screenshot 2023-08-30 at 15 29 22

**Edit:** Ok, it's expected as it comes from the custom action included in testing instructions 😅 That would mean everything works correctly.

The same issue with the nested Query Loop exists on the post page though as in the block theme.

Screenshot 2023-08-30 at 15 36 32

The homepage on the front doesn't render the nested query loop for the thirs post, too:

Screenshot 2023-08-30 at 15 37 14

It sounds reasonable approach, but differs from how it is handled in the block theme. Maybe it uses the post excerpt though? 😅

**Comment by felixarntz** [comment]
Thank you for the detailed testing @gziolo! Some follow up feedback on your notes:
* I think your observations on the "third post with loop" aren't quite right: on the frontend, when viewed as the singular post in TT3 theme, its query loop only displays itself, which is correct, since the global query loop on that URL is only that post. I don't think there's any real use-case for doing that anyway, but it's working as it should. If anything, that part is off in the editor, where it should also display just that post, not the list of all posts. It should only display all the posts when looking at the post in an archive. Basically what is displayed in that post will differ depending on the current main query loop.
* Regarding your observation in the TT1 theme, as you've already suspected, in the archive view the loop inside "third post with loop" is not displayed as that template is only using the excerpt where any "rich content" is cut.

**Review by gziolo — APPROVED** [review]
@felixarntz, appreciate the detailed explanation to what I’ve incorrectly reported. I believe, it warrants at least a follow up issue in the Gutenberg repo to consider some UI improvements in the editor.

**Comment by felixarntz** [comment]
Thanks @gziolo. Based on the findings, I also left an update on https://github.com/WordPress/gutenberg/issues/50790: I don't think it makes sense that we even allow inheriting the main query inside a specific post.

**Review by mukeshpanchal27 — APPROVED** [review]
Thanks @felixarntz for PR. 

https://github.com/WordPress/wordpress-develop/pull/5104#discussion_r1309421673 LGTM.   When you apply those changes please combine `global` variables.
```
global $_wp_current_template_content, $wp_embed, $wp_query;
```

File: src/wp-includes/block-template.php (hunk 3 above)
Comment (joemcgill) [reply src/wp-includes/block-template.php]: I do think this is probably safer, but pretty unclear what's going on here unless someone does a git blame and finds this conversation (👋🏻 hi future person, if so!). Can you add an inline comment explaining why this is here and why you need to check the `post_count`?

**Comment by joemcgill** [comment]
A new concern I have after thinking about this some more, is that it essentially forces all singular block templates to include the entire template in "the loop", which can cause inconsistent behavior for things like headers or other template parts that are traditionally outside the loop. For example, if we have some logic that attempts to apply some image optimizations only when the image is inside the loop, with this change, a header image would be included on singular pages but not on list pages like archives. This is very edge case, but could be confusing. Would it be safer to only apply this logic to the full template if we are able to first do a `has_block()` check to see if a theme has omitted one of the blocks that would start the loop as expected?

**Comment by felixarntz** [comment]
@joemcgill Your concern is a fair point, and I've considered that initially when I started this PR.

> Would it be safer to only apply this logic to the full template if we are able to first do a `has_block()` check to see if a theme has omitted one of the blocks that would start the loop as expected?

I did have that in here at the start (also see https://github.com/WordPress/wordpress-develop/pull/5104#discussion_r1309098376), but there are several limitations with that:
* There is no existing core block that would realistically be used on a singular template to trigger the main query loop. I thought originally that the `core/query` and `core/post-template` loop would also be usable for that purpose, but they unfortunately include additional markup _forcing_ posts to be in a semantic list, which makes no sense for singular content and can be problematic for accessibility.
    * Figuring this out would require notably more discussion and changes. For example, would we introduce a new block for singular loops? Or somehow adjust the `core/post-template` block to conditionally not have wrapping markup? Those things may end up quite confusing from an end user perspective too.
* Even if there were singular-friendly variant of these blocks, we would need something more complex than `has_block()`, because we would only _not_ need to trigger the main query if the `core/post-template` block was wrapped in a `core/query` block that had the `inherit: true` attribute set - something more costly to check for which there's no dedicated function available.

While what you're raising is indeed a potential concern, I see it more as a theoretical concern. While typically themes indeed only have the loop around the main content of a singular URL, I can't think of any actual problem that wrapping the whole content of such a URL would cause.

In other words: I think this is something worth revisiting in the future, but I don't consider it a blocker, since the current issue is certainly more severe, and properly fixing the additional concern would require more discussion and effort than what we could realistically tackle for 6.4.

File: src/wp-includes/block-template.php (hunk 3 above)
Comment (felixarntz) [reply src/wp-includes/block-template.php]: @joemcgill Updated in https://github.com/WordPress/wordpress-develop/pull/5104/commits/766d453fbff73a99391f4f512126654f6e796bc2.

**Review by costdev — APPROVED** [review]
Thanks for the PR @felixarntz! And for adding test coverage after our discussion in Slack. This looks good to me.

If the edge case mentioned above is a very rare and minor one, then I agree we should address it as soon as possible, and that it shouldn't be a blocker for this PR.

**Comment by felixarntz** [comment]
Committed in https://core.trac.wordpress.org/changeset/56507
